src/libstd/ffi/c_str.rs

```diff
@@ -260,8 +267,14 @@ impl CString {
     #[stable(feature = "cstr_memory", since = "1.4.0")]
     pub unsafe fn from_raw(ptr: *mut c_char) -> CString {
         let len = libc::strlen(ptr) + 1; // Including the NUL byte
-        let slice = slice::from_raw_parts(ptr, len as usize);
-        CString { inner: mem::transmute(slice) }
+        if len == 1 {
+            debug_assert_eq!(ptr, NUL_TERMINATED_EMPTY.as_ptr() as *mut c_char);
```

Comment:
This assertion makes no sense. It is requiring all `ptr`s pointing to null byte be `NUL_TERMINATED_EMPTY`, which is a breaking change and also weird.

(It precludes code like one below from working)

```rust
let cstring = CString::from_raw(b"\0");
...
mem::forget(cstring);
```